node/src/testing/test_block_builder.rs

```diff
@@ -0,0 +1,194 @@
+use std::collections::BTreeMap;
+
+use casper_hashing::Digest;
+use casper_types::{
+    system::auction::BLOCK_REWARD, EraId, ProtocolVersion, PublicKey, SecretKey, Timestamp, U512,
+};
+use rand::Rng;
+
+use crate::{
+    components::consensus::EraReport,
+    types::{Block, BlockHash, BlockPayload, Deploy, DeployHashWithApprovals, FinalizedBlock},
+};
+
+const MAX_ERA_FOR_RANDOM_BLOCK: u64 = 6;
+
+pub(crate) struct TestBlockBuilder {
+    era: Option<u64>,
+    height: Option<u64>,
+    protocol_version: ProtocolVersion,
+    is_switch: Option<bool>,
+    deploys: Vec<Deploy>,
+    state_root_hash: Option<Digest>,
+    parent_hash: Option<BlockHash>,
+    timestamp: Option<Timestamp>,
+}
+
+impl TestBlockBuilder {
+    pub(crate) fn new() -> Self {
+        Self {
+            era: None,
+            height: None,
+            protocol_version: ProtocolVersion::V1_0_0,
+            is_switch: None,
+            deploys: Vec::new(),
+            state_root_hash: None,
+            parent_hash: None,
+            timestamp: None,
+        }
+    }
+
+    pub(crate) fn era(mut self, era: u64) -> Self {
+        self.era = Some(era);
+        self
+    }
+
+    #[allow(unused)]
+    pub(crate) fn height(mut self, height: u64) -> Self {
+        self.height = Some(height);
+        self
+    }
+
+    #[allow(unused)]
+    pub(crate) fn switch_block(mut self, is_switch: bool) -> Self {
+        self.is_switch = Some(is_switch);
+        self
+    }
+
+    #[allow(unused)]
+    pub(crate) fn state_root_hash(mut self, state_root_hash: Digest) -> Self {
+        self.state_root_hash = Some(state_root_hash);
+        self
+    }
+
+    #[allow(unused)]
+    pub(crate) fn parent_hash(mut self, parent_hash: BlockHash) -> Self {
+        self.parent_hash = Some(parent_hash);
+        self
+    }
+
+    pub(crate) fn deploys<'a, I: IntoIterator<Item = &'a Deploy>>(
+        mut self,
+        deploys_iter: I,
+    ) -> Self {
+        self.deploys = deploys_iter.into_iter().cloned().collect();
+        self
+    }
+
+    #[allow(unused)]
+    pub(crate) fn timestamp(mut self, timestamp: Timestamp) -> Self {
+        self.timestamp = Some(timestamp);
+        self
+    }
+
+    pub(crate) fn build<R: Rng + ?Sized>(self, rng: &mut R) -> Block {
```

Comment:
Again, I'd prefer if this took a `TestRng` rather than `R: Rng`.

Reply:
Changed in https://github.com/casper-network/casper-node/pull/3816/commits/2b0eefcbeb36d4742f2b0fc35b4e36711d4a5f25